### PR TITLE
Fix custom allreduce warmup values

### PR DESCRIPTION
**Pull request by 1sh1ro**
# Fix CustomAllreduce warmup to return valid reduced values

## Summary

During CUDA graph setup, `CustomAllreduce.custom_all_reduce()` currently returns
`torch.empty_like(input)` for warmup forwards that run while `_IS_CAPTURING` is
set but the current stream is not actively capturing.

That preserves the out-of-place allocation pattern, but it can return
uninitialized values to the model forward. Some model paths can consume warmup
outputs or update persistent runtime state such as KV cache or indexing buffers
during warmup. In those cases, returning uninitialized data can poison later
hidden states and surface as non-finite logits.

This PR changes the warmup branch to run the normal non-registered allreduce:

```python
return self.all_reduce(input, registered=False)
```

The actual stream-capture path still uses `registered=True`, so the steady-state
CUDA graph replay fast path is unchanged.

## Motivation

Warmup forwards should be allowed to exercise the same numerical data path as
normal forwards. Returning an uninitialized tensor is only safe if the caller
never consumes that value and the forward has no persistent side effects.

That assumption is fragile for graph setup paths where model forwards may
populate runtime state during warmup. In those cases, an uninitialized warmup
output can propagate non-finite values into later execution.

Using a valid allreduce result during warmup fixes the issue while keeping
custom allreduce and CUDA graph replay enabled.

## Duplicate check

I searched open PRs for:

- `custom allreduce warmup`
- `CustomAllreduce empty_like`
- `custom_all_reduce CUDA graph warmup`

I did not find an existing open PR that changes this warmup branch to return a
valid allreduce result.

## Testing

- Added a unit test covering the warmup branch and verifying it calls
  `all_reduce(..., registered=False)` and returns that reduced value.
- Locally ran:

```text
python3 -m compileall -q \
  vllm/distributed/device_communicators/custom_all_reduce.py \
  tests/distributed/test_custom_all_reduce.py
```

`pytest` and `ruff` were not installed in this local environment, so I could not
run the full test suite here.

## AI assistance

This PR was prepared with AI assistance. The changed lines and test coverage
were reviewed before submission.

### CODE REVIEW

**Comment by 1sh1ro** [comment]
Could a maintainer please review this change and add the `ready` label if it looks appropriate? The PR is a targeted bug fix for `CustomAllreduce` warmup behavior, includes a unit test, and the duplicate-work check has been added to the description.
